docs/slightfile-design.md

```diff
@@ -0,0 +1,48 @@
+# `slightfile` Re-Design
+
+## Problem
+
+As of 2022-09-15, the `slightfile` design has two big restrictions:
+(1) we do not allow for multiple-capabilities of the same namespace but different implementor types (e.g., `kv.azblob`, and `kv.filesystem`), and
```

Comment:
@Mossaka what do you mean when you say we "should consider if it makes sense to declare the required configuration variables in `slightfile`"?

Reply:
I was thinking about something like this when I said declaring required configuration variables in slightfile

```toml
[[config]]
name = "azure-blob-config"
storage_account = "{AZURE_STORAGE_ACCOUNT}"
storage_password = "{AZURE_STORAGE_PASSWORD}"
```

It would make more sense to use the world "binding" to environment variables, than "declare".

Reply:
I like how explicit the declarative config is. What about something like:

```toml
[[config]]
name = "azure-blob-config"
storage_account = "{env.AZURE_STORAGE_ACCOUNT}"
storage_password = "{my_keyvault.AZURE_STORAGE_PASSWORD}"
```

Where we are able to express the binding such that we can be clear where the variable is sourced from.

Reply:
In the future, here's how I'm picturing we can have this interaction:

@ ~/repos
```sh
> slight new --rust spidey && cd spidey
```

@ ~/repos/spidey
```sh
> tree -L 1
├── Cargo.toml
├── slightfile.toml
└── src
```

@ ~/repos/spidey
```sh
> cat slightfile.toml
specversion = "0.1"
secret_store = "configs.envvars"
```

@ ~/repos/spidey
```sh
> slight add kv.azblob && cat slightfile.toml
specversion = "0.1"
secret_store = "configs.envvars"

[[capability]]
name = "kv.azblob"
ref = ""
configs = "" # if absent, defaults to global "secret_store"
AZURE_STORAGE_ACCOUNT = ""
AZURE_STORAGE_KEY = ""
```

The `add` command could also auto-populate the fields mentioned in your second bullet-point, @Mossaka.

Reply:
> Note: The `AZURE_STORAGE_ACCOUNT` and `AZURE_STORAGE_KEY` would not contain the values of the config, but the name they are stored w/ under a specific config type.

Reply:
I like how you pictured this in the interaction with `slight` CLI. This command is very convenient for users! The reason I proposed a new array of objects `[[config]]` is to make it sharable. Imagine you have a `azblob` config, with storage account and key specified. You can open a few instances of this capability, with different namespaces:

```rust
kv::open("orders");
kv::open("waitlist");
```

You can specify both "orders" and "waitlist" capability use the same shared config object in slightfile.

Reply:
That's a very good point. I like this declarative shareable aspect of configs — will update the doc.